### PR TITLE
More Popup focus fixes

### LINKED ISSUE #4808: Focus is lost after selecting an item on ComboBox
.NET Core 3.1.8
Avalonia 0.10.0-preview5

XAML:
```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="test_avalonia.MainWindow"
        Title="test_avalonia">
    <StackPanel>
        <Button Content="Button"/>
        <ComboBox>
            <ComboBoxItem Content="Item1"/>
            <ComboBoxItem Content="Item2"/>
            <ComboBoxItem Content="Item3"/>
            <ComboBoxItem Content="Item4"/>
            <ComboBoxItem Content="Item5"/>
        </ComboBox>
    </StackPanel>
</Window>
```

At start, the focus is on MainWindow, and I can cycle through controls with Tab key, as expected.
image

After selecting an item on ComboBox, with either mouse or keyboard, the focus is lost.
image

On 0.9.12, it works as expected, the focus is still in the window and I can tab through controls as normal.

### PR DESCRIPTION
**Pull request by amwx**
## What does the pull request do?

My last Popup PR allowed focus to be directed into a Popup, but Popup doesn't return focus out of the popup when it closes leading to the current focus being set to null and no way to tab navigate after closing.

My solution is to set Focus on the PlacementTarget of the popup upon closing (simplest solution). Alternatively, we could store the current focus on open and restore it on close (this should be the PlacementTarget anyway). Not sure which is better or if there's a better solution, but I'm open to suggestions. I also wasn't sure if PlacementTarget can be null (from Open() it looks like it can be), so there's some logic to try to set focus in that case too, just in case.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Fixed issues
Fixes #4808

### CODE REVIEW

File: src/Avalonia.Controls/Primitives/Popup.cs [hunk 1]
```diff
@@ -586,6 +586,19 @@ private void CloseCore()
...
 
             Closed?.Invoke(this, EventArgs.Empty);
+
+            if(PlacementTarget != null)
+            {
+                FocusManager.Instance?.Focus(PlacementTarget);
+            }
+            else
+            {
+                var anc = this.GetLogicalAncestors().OfType<IControl>().FirstOrDefault();
```
Comment (MarchingCube) [inline src/Avalonia.Controls/Primitives/Popup.cs:596]:
```suggestion
                var anc = this.FindLogicalAncestorOfType<IControl>();
```

File: src/Avalonia.Controls/Primitives/Popup.cs [hunk 2]
```diff
@@ -586,6 +586,19 @@ private void CloseCore()
             }
 
             Closed?.Invoke(this, EventArgs.Empty);
+
+            if(PlacementTarget != null)
```
Comment (MarchingCube) [inline src/Avalonia.Controls/Primitives/Popup.cs:590]:
```suggestion
            if (PlacementTarget != null)
```

**Review by MarchingCube — DISMISSED** [review]

File: src/Avalonia.Controls/Primitives/Popup.cs (hunk 2 above)
Comment (amwx) [reply src/Avalonia.Controls/Primitives/Popup.cs]: Fixed

File: src/Avalonia.Controls/Primitives/Popup.cs (hunk 1 above)
Comment (amwx) [reply src/Avalonia.Controls/Primitives/Popup.cs]: Fixed. I also fixed this in `Open()` where I originially copied this from

**Comment by maxkatz6** [comment]
What if I close popup with light dismiss by clicking on some TextBox outside of it?
I would expect TextBox to be focused in that case, and not popup's PlacementTarget.
Probably we need to check if focus were inside of popup before it closed.

**Comment by amwx** [comment]
> 
> 
> What if I close popup with light dismiss by clicking on some TextBox outside of it?
> I would expect TextBox to be focused in that case, and not popup's PlacementTarget.
> Probably we need to check if focus were inside of popup before it closed.

Comparing to WinUI, I'm getting the same behavior. If focus is in a Popup and I click on a textbox (or something) outside of the popup, the PlacementTarget is still focused on close and it takes two clicks to focus the outside textbox. Ex I did in the WinUI xaml controls gallery, launch a datepicker and tab to ensure focus is on one of the pickers in the popup, then click on the search box in the NavView. The popup closes, the datepicker is focused and it takes a second click to focus the search box.
IIRC the light dismiss layer logic absorbs the click which is why this happens.

**Comment by maxkatz6** [comment]
> IIRC the light dismiss layer logic absorbs the click which is why this happens
That's true, I provided bad example.
Need to check case, when light dismiss is disabled (so user can focus control outside of popup without closing it), and close popup programmatically.

**Comment by amwx** [comment]
> Need to check case, when light dismiss is disabled (so user can focus control outside of popup without closing it), and close popup programmatically.

Ok I see what you're saying. I'll take a look.

**Comment by amwx** [comment]
I've added a check. In closing the popup, focus gets set to null if the popup had it, so I just check for that and only set the focus on the PlacementTarget if current focus is null. In case of `IsLightDismissEnabled=false`, this prevents moving the focus if focus is set on another control e.g. TextBox and popup is programmatically closed.

**Comment by maxkatz6** [comment]
@amwx thanks.
Can we have some tests just for case?

**Review by maxkatz6 — APPROVED** [review]
LGTM

**Pull request merged**